### PR TITLE
WIP: Cache/pre-render refactoring

### PR DESCRIPTION
**Pull request by fnevgeny**
OK, here we go.

In spite of the original intentions in #12, #211, I stick to the essentially single-thread model. With the changes detailed below, it seems to perform very well. There are also a bunch of fine-tuning knobs to fit better for specific needs/hardware limitations, if desired.

- First, cache is _always_ present. At the minimum, it is used for the "overview" thumbnails. The rationale is that without the images, overview is largely useless - it's faster just to enter the desired slide number with `g`. The rendering is performed at the time the overview is shown for the first time - and only for the thumbnails visible in the scrolled list. So it should be reasonably fast. And people who don't use this feature won't notice any impact at all.

- Pre-rendering is done only for a limited number of slides (adjustable; default is 2) ahead of the current position. For `View`'s that show only "complete" user slides (i.e., "next") this means 2 _user_ slides, for others - real slides. 0 disables pre-rendering, -1 would pre-render all at once (the "old" behavior, if somebody is so inclined...).

- The pre-rendering is triggered by page changes, but is slightly delayed (4 sec; adjustable). This is done to avoid excessive rendering load if slides are "fast forwarded" looking for something.

- Once in a while (60 sec; adjustable) a cache cleaner is fired which removes expired (not _accessed_ during the last 10 min; adjustable) cached images. There are exceptions, though, in particular, the overview thumbnails, which are kept permanently. Another exception is slides that took a significant CPU time (1000 ms; adjustable) to render - trading memory for speed, even if the chance it will be needed again is low.

- Another adjustment in the speed/memory balance is the (uncompressed) size. Small images (256 kB by default) are kept as is. This also increases the speed of the overview rendering.

- On fast hardware one might want to disable pre-rendering but still use cache (in addition to the "overview" mini slides which are always cached, that is). Then (fast hardware) it would make sense to bother caching only relatively "heavy" slides. This threshold is also adjustable, as you might guess (10 ms by default).

- Finally, there is `cache-debug` boolean option to enable printerr()'ing some info at run time. This will likely go away before the release.

What do you think? On the entire approach, defaults, ...?

### CODE REVIEW

File: man/pdfpcrc.in [hunk 1]
```diff
@@ -66,20 +66,30 @@ The following option_name's are recognized:
...
 Add an additional black slide at the end of the presentation (bool, Default is false).
 .TP
+.BI cache-clean-period
+Period (in units of s) to check for expired cached content (int, Default is 60).
+.TP
+.BI cache-expiration
+Time (in units of s) to keep (pre)rendered slides in the cache (int, Default is
+600). Set to a negative value to keep cached slides permanently.
+.TP
+.BI cache-max-rtime
```
Comment (AndreasBilke) [inline man/pdfpcrc.in]: The default values are missing?

Comment (fnevgeny) [reply man/pdfpcrc.in]: Oops, fixed.

**Comment by fnevgeny** [comment]
> The only thing I am not sure is weather we need that compressed/uncompressed distinction (for complexity reasons). Probably it's enough to store any image compressed in memory.

I wanted to provide possibility for backward compatibility (to imitate the former `-z` behavior). With that in place, supporting also anything in between always/never is a minor issue. 

> If a computer is so slow that the compression is a real timing problem, the user could increase the number of pre rendered slides.

But if the computer is slow, pre-rendering many slides at once will make pdfpc laggish at the startup - we are back to square one (or more precisely, #12 :)).

At some point I was thinking about adding a separate low-priority worker executed periodically to compress caches (which would initially be all stored uncompressed). Maybe it's worth doing?

**Comment by AndreasBilke** [comment]
> I wanted to provide possibility for backward compatibility (to imitate
> the former -z behavior). With that in place, supporting also anything in
> between always/never is a minor issue.

But v4.4 will be a breaking change anyways :-)

> At some point I was thinking about adding a separate low-priority worker
> executed periodically to compress caches (which would initially be all
> stored uncompressed). Maybe it's worth doing?

I don't know. Do you now a computer or slide set which is unusable slow?
Probably it is to much effort for a tiny (tiny) fraction of users/setups.

**Comment by einhander** [comment]
Hi, I occasionally use an old computer which is very slow without -z option. And I saw one presentation which fails to load  without -z option.

**Comment by fnevgeny** [comment]
> Hi, I occasionally use an old computer which is very slow without -z option. And I saw one presentation which fails to load without -z option.

@einhander, can you please test this branch with that presentation/computer? Also, setting `option cache-max-usize` first to -1 (equivalent of the previous `-z`) and then to 0 (always compress).

**Comment by fnevgeny** [comment]
> But v4.4 will be a breaking change anyways :-)

Right, but I wasn't talking about the backward compatibility of the command-line flags (which is actually broken), but the ability to disable the compression at all.

> I don't know. Do you now a computer or slide set which is unusable slow?

No, I'm unhappy to have rather upper-end hardware at home and work :-). But wait, I must have an old Raspberry PI somewhere collecting dust. Wouldn't it be nice to carry your presentation box in your pocket? :)

**Comment by einhander** [comment]
@fnevgeny I can test it, only at mid of September. 
+1 for a Raspberry PI, it is really slow)))

**Comment by AndreasBilke** [comment]
@fnevgeny What is missing in your PR to become mergable?

**Comment by fnevgeny** [comment]
> What is missing in your PR to become mergable?

Well, it needs a minor rebasing, but it's trivial.

During the reworking of the history functionality which was a must for the cache refactoring, among other things, I intentionally broke the `previous_seen()` and `next_unseen()` actions. First, I have a feeling that the (properly working) history navigation solves at least part of the problems these actions were supposed to deal with. Second, if these must remain, I'd rather implement them differently, without an extra array - which is a potential source of crashes. This can be done still within this separate branch, or after merging.

Another history-related question is whether we need the forward history navigation as well. Again, this can be done before or after merging.

I've actually been using this branch for all my own presentations lately, and haven't encountered any problems.

**Comment by AndreasBilke** [comment]
> First, I have a feeling that the (properly working) history navigation
> solves at least part of the problems these actions were supposed to deal
> with. Second, if these must remain, I'd rather implement them
> differently, without an extra array - which is a potential source of
> crashes.

I'm not sure if I understand you correctly: Do you want to keep this
functionality (and make it right) or do you want to purge it?

> Another history-related question is whether we need the forward history
> navigation as well.

Same question as before :)

**Comment by fnevgeny** [comment]
> I'm not sure if I understand you correctly: Do you want to keep this functionality (and make it right) or do you want to purge it?

IMO, its utility is in very specific corner use cases if the history navigation is properly working (including forward one), so I'd vote for its removal. However, it might be wise to hear opinions of those who use it frequently - or at least asked to be implemented in the first case (#242 ).

>> Another history-related question is whether we need the forward history navigation as well.

> Same question as before :)

I think it would be nice to have, it seems to me quite straightforward to code.

**Comment by AndreasBilke** [comment]
> IMO, its utility is in very specific corner use cases if the history
> navigation is properly working (including forward one), so I'd vote for
> its removal. However, it might be wise to hear opinions of those who use
> it frequently - or at least asked to be implemented in the first case
> (#242).

I can assume what the orginal requestor will say about :) If I had to
decide between less code (less complex) and a feature for a specific corner
case I would choose less code.

**Comment by fnevgeny** [comment]
> I can assume what the orginal requestor will say about :) If I had to decide between less code (less complex) and a feature for a specific corner case I would choose less code.

OK, then I suggest merging the branch (I've just rebased it against master so there should be no conflicts). I will follow with extending the history for forward navigation. Regarding the dropped functionality, I promise to reconsider it if there is a strong demand.

**Pull request merged**